shared/ssz/hash_cache_test.go

```diff
@@ -0,0 +1,216 @@
+		Hash: common.HexToHash("0x0123456"),
+	}
+
+	key, err := hashKeyFn(mRoot)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if key != mRoot.Hash.Hex() {
+		t.Errorf("Incorrect hash key: %s, expected %s", key, mRoot.Hash.Hex())
+	}
+}
+
+func TestHashKeyFn_InvalidObj(t *testing.T) {
+	_, err := hashKeyFn("bad")
+	if err != ErrNotMarakleRoot {
+		t.Errorf("Expected error %v, got %v", ErrNotMarakleRoot, err)
+	}
+}
+
+func TestObjCache_byHash(t *testing.T) {
+	cache := newHashCache()
+
+	byteSl := [][]byte{{0, 0}, {1, 1}}
+	mr, err := merkleHash(byteSl)
+	if err != nil {
+		t.Fatal(err)
+	}
+	hs, err := hashedEncoding(reflect.ValueOf(byteSl))
+	if err != nil {
+		t.Fatal(err)
+	}
+	exists, _, err := cache.RootByHash(bytesutil.ToBytes32(hs))
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	if exists {
+		t.Error("Expected block info not to exist in empty cache")
+	}
+
+	if _, err := cache.AddRetriveMarkleRoot(byteSl); err != nil {
+		t.Fatal(err)
+	}
+
+	exists, fetchedInfo, err := cache.RootByHash(bytesutil.ToBytes32(hs))
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !exists {
+		t.Error("Expected blockInfo to exist")
+	}
+	if bytes.Compare(mr, fetchedInfo.MarkleRoot) != 0 {
+		t.Errorf(
+			"Expected fetched info number to be %v, got %v",
+			mr,
+			fetchedInfo.MarkleRoot,
+		)
+	}
+	if fetchedInfo.Hash != bytesutil.ToBytes32(hs) {
+		t.Errorf(
+			"Expected fetched info hash to be %v, got %v",
+			hs,
+			fetchedInfo.Hash,
+		)
+	}
+}
+
+func TestMerkleHashWithCache(t *testing.T) {
+	cache := newHashCache()
+	for i := 0; i < 200; i++ {
+
+		runMerkleHashTests(t, func(val [][]byte) ([]byte, error) {
+			return merkleHash(val)
+		})
+
+	}
+
+	for i := 0; i < 200; i++ {
+
+		runMerkleHashTests(t, func(val [][]byte) ([]byte, error) {
+			return cache.AddRetriveMarkleRoot(val)
+		})
+
+	}
+
+}
+
+// EncodeDepositData converts a deposit input proto into an a byte slice
+// of Simple Serialized deposit input followed by 8 bytes for a deposit value
+// and 8 bytes for a unix timestamp, all in LittleEndian format.
+func EncodeDepositData(
+	depositInput *pb.DepositInput,
+	depositValue uint64,
+	depositTimestamp int64,
+) ([]byte, error) {
+	wBuf := new(bytes.Buffer)
+	if err := Encode(wBuf, depositInput); err != nil {
+		return nil, fmt.Errorf("failed to encode deposit input: %v", err)
+	}
+	encodedInput := wBuf.Bytes()
+	depositData := make([]byte, 0, 512)
+	value := make([]byte, 8)
+	binary.LittleEndian.PutUint64(value, depositValue)
+	timestamp := make([]byte, 8)
+	binary.LittleEndian.PutUint64(timestamp, uint64(depositTimestamp))
+
+	depositData = append(depositData, value...)
+	depositData = append(depositData, timestamp...)
+	depositData = append(depositData, encodedInput...)
+
+	return depositData, nil
+}
+
+// generateInitialSimulatedDeposits generates initial deposits for creating a beacon state in the simulated
+// backend based on the yaml configuration.
+func generateInitialSimulatedDeposits(numDeposits uint64) ([]*pb.Deposit, []*bls.SecretKey, error) {
```

Comment:
Again this makes no sense, why are we adding a function from the simulated backend here.

Reply:
couldn't import simulated backend because of circular dependency so that is what I chose to do in order to generate data that is similar for what we are going to process

Reply:
imo this isnt a good place for this, the ssz package is a shared package which might be used by other projects if they want to use ssz. Bringing eth 2.0 specific logic here isnt good, we want to keep it as self contained as possible. If you want to test how we tree hash proto files, you can just create a proto object with junk data. There is no reason to use simulated backend here, for what is a serialization library.

Reply:
removed